component/loki/source/aws_firehose/internal/handler.go

```diff
@@ -0,0 +1,255 @@
+		return
+	}
+
+	// common labels contains all record-wide labels
+	commonLabels := labels.NewBuilder(nil)
+	commonLabels.Set("__aws_firehose_request_id", req.Header.Get("X-Amz-Firehose-Request-Id"))
+	commonLabels.Set("__aws_firehose_source_arn", req.Header.Get("X-Amz-Firehose-Source-Arn"))
+
+	// if present, use the tenantID header
+	tenantHeader := req.Header.Get("X-Scope-OrgID")
+	if tenantHeader != "" {
+		commonLabels.Set(lokiClient.ReservedLabelTenantID, tenantHeader)
+	}
+
+	h.metrics.batchSize.WithLabelValues().Observe(float64(len(firehoseReq.Records)))
+
+	for _, rec := range firehoseReq.Records {
+		// cleanup err since it might have failed in the previous iteration
+		err = nil
+
+		decodedRecord, recordType, err := h.decodeRecord(rec.Data)
+		if err != nil {
+			h.metrics.errorsRecord.WithLabelValues(getReason(err)).Inc()
+			level.Error(h.logger).Log("msg", "failed to decode request record", "err", err.Error())
+			continue
+		}
+
+		h.metrics.recordsReceived.WithLabelValues(string(recordType)).Inc()
+
+		// todo(pablo): add use incoming timestamp option
+
+		switch recordType {
+		case OriginDirectPUT:
+			h.sender.Send(req.Context(), loki.Entry{
+				Labels: h.postProcessLabels(commonLabels.Labels(nil)),
+				Entry: logproto.Entry{
+					Timestamp: time.Now(),
+					Line:      string(decodedRecord),
+				},
+			})
+		case OriginCloudwatchLogs:
+			err = h.handleCloudwatchLogsRecord(req.Context(), decodedRecord, commonLabels.Labels(nil))
+		}
+		if err != nil {
+			h.metrics.errorsRecord.WithLabelValues(getReason(err)).Inc()
+			level.Error(h.logger).Log("msg", "failed to handle cloudwatch record", "err", err.Error())
+			continue
+		}
+	}
+
+	sendAPIResponse(w, firehoseReq.RequestID, "", http.StatusOK)
+}
+
+// postProcessLabels applies relabels, then drops not relabeled internal and invalid labels.
+func (h *Handler) postProcessLabels(lbs labels.Labels) model.LabelSet {
+	// apply relabel rules if any
+	if len(h.relabelRules) > 0 {
+		lbs, _ = relabel.Process(lbs, h.relabelRules...)
+	}
+
+	entryLabels := make(model.LabelSet)
+	for _, lbl := range lbs {
+		// if internal label and not reserved, drop
+		if strings.HasPrefix(lbl.Name, "__") && lbl.Name != lokiClient.ReservedLabelTenantID {
+			continue
+		}
+
+		// ignore invalid labels
+		if !model.LabelName(lbl.Name).IsValid() || !model.LabelValue(lbl.Value).IsValid() {
+			continue
+		}
+
+		entryLabels[model.LabelName(lbl.Name)] = model.LabelValue(lbl.Value)
+	}
+	return entryLabels
+}
+
+// sendAPIResponse responds to AWS Firehose API in the expected response format. To simplify error handling,
+// it uses a string template instead of marshalling a struct.
+func sendAPIResponse(w http.ResponseWriter, firehoseID, errMsg string, status int) {
+	timestamp := time.Now().Unix()
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(status)
+	if errMsg != "" {
+		_, _ = w.Write([]byte(fmt.Sprintf(errorResponseTemplate, firehoseID, timestamp, errMsg)))
```

Comment:
Does AWS firehose expect that we respond with http.StatusOK even if an error occurred?